Provide integration with Diactoros FilterServerRequestInterface

weierophinney opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

This patch adds support for the `FilterServerRequestInterface` in the `ServerRequestFactoryFactory`.
When the `FilterServerRequestInterface` service is defined, it will be used for the `$requestFilter` argument of Diactoros' `ServerRequestFactory::fromGlobals` factory.

Additionally, the patch adds a specialized factory for configuring a `FilterUsingXForwardedHeaders` to use for this service.
Users can configure it using the following configuration scheme:

```php
$config = [
    'laminas-diactoros' => [
        'server-request-filter' => [
            'x-forwarded-headers' => [
                'trusted-proxies' => string|string[],
                'trusted-headers' => string[],
            ],
        ],
    ],
    'dependencies' => [
        'factories' => [
            \Laminas\Diactoros\ServerRequestFilter\FilterServerRequestInterface::class =>
                \Mezzio\Container\FilterUsingXForwardedHeadersFactory::class,
        ],
    ],
];
```

If they want to opt-out of usage of `X-Forwarded-*` headers entirely, they can now do so with the following configuration:

```php
$config = [
    'dependencies' => [
        'aliases' => [
            \Laminas\Diactoros\ServerRequestFilter\FilterServerRequestInterface::class =>
                \Laminas\Diactoros\ServerRequestFilter\DoNotFilter::class,
        ],
    ],
];
```

weierophinney committed:
feat: add default definition for Diactoros ServerRequestFilterInterface service

This patch adds support for the Diactoros ServerRequestFilterInterface service via the following mechanisms:

- It maps it to the `XForwardedHeaderFilter` factory.
- It provides default configuration for the `XForwardedHeaderFilter` factory that trusts any proxy by default (existing behavior).

Users can provide an alternate mapping for the `ServerRequestFilterInterface` or provide alternate configuration for the `XForwardedHeaderFilter` factory if they want to remove support for trusting proxy headers or provide explicit proxies and/or headers to trust.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: adds support for Diactoros FilterServerRequestInterface

This patch adds support for the `FilterServerRequestInterface` in the `ServerRequestFactoryFactory`.
When the `FilterServerRequestInterface` service is defined, it will be used for the `$requestFilter` argument of Diactoros' `ServerRequestFactory::fromGlobals` factory.

Additionally, the patch adds a specialized factory for configuring a `FilterUsingXForwardedHeaders` to use for this service.
Users can configure it using the following configuration scheme:

```php
$config = [
    'laminas-diactoros' => [
        'server-request-filter' => [
            'x-forwarded-headers' => [
                'trusted-proxies' => string|string[],
                'trusted-headers' => string[],
            ],
        ],
    ],
    'dependencies' => [
        'factories' => [
            \Laminas\Diactoros\ServerRequestFilter\ServerRequestFilterInterface::class =>
                \Mezzio\Container\FilterUsingXForwardedHeadersFactory::class,
        ],
    ],
];
```

If they want to opt-out of usage of `X-Forwarded-*` headers entirely, they can now do so with the following configuration:

```php
$config = [
    'dependencies' => [
        'aliases' => [
            \Laminas\Diactoros\ServerRequestFilter\ServerRequestFilterInterface::class =>
                \Laminas\Diactoros\ServerRequestFilter\DoNotFilter::class,
        ],
    ],
];
```

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix CS issues as flagged by phpcs

Unneeded import.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on docs/book/v3/features/container/factories.md:
```diff
@@ -250,9 +250,25 @@ As an example:
+'dependencies' => [
+    'aliases' => [
+        Laminas\Diactoros\ServerRequestFilter\FilterServerRequestInterface::class => 
+            Laminas\Diactoros\ServerRequestFilter\DoNotFilter::class,
```
Careful: the `DoNotFilter` is not a declared service, so an alias does not suffice.

weierophinney replied:
Arg, this should be `invokables`; I'll resolve.

weierophinney replied:
Done.

Ocramius commented on docs/book/v3/features/container/factories.md:
```diff
@@ -510,3 +526,42 @@ When creating the `PhpRenderer` instance, it will inject it with a
+    'laminas-diactoros' => [
+        'server-request-filter' => [
+            'x-forwarded-headers' => [
+                'trusted-proxies' => string|string[],
```
argh, `string|string[]` survived? @weierophinney let's please keep signatures small in future - if a union type is not **necessary**, it shouldn't be added.

Too late for that here.

weierophinney replied:
No - this is specific to the factory implementation in this package.

I can require `list<string>` here, though.

Ocramius replied:
Let's simplify to the bone then - allowing multiple types doesn't bring us much

weierophinney replied:
Done.

Ocramius commented on src/Container/FilterUsingXForwardedHeadersFactory.php:
```diff
@@ -0,0 +1,78 @@
+ * 'laminas-diactoros' => [
+ *     'server-request-filter' => [
+ *         'x-forwarded-headers' => [
+ *             // Trust any proxy:
+ *             'trusted-proxies' => '*',
```
Let's get rid of this example/scenario - simplify signatures to a minimum, where possible.

weierophinney replied:
Not sure what you're getting at - are you suggesting `['*']` as the value here, or something else?

Ocramius replied:
Correct, `['*']` would be more appropriate :+1:

weierophinney replied:
Done.

Ocramius commented on src/Container/FilterUsingXForwardedHeadersFactory.php:
```diff
@@ -0,0 +1,78 @@
+        $config = $container->get('config');
+        $config = $config[ConfigProvider::DIACTOROS_CONFIG_KEY][ConfigProvider::DIACTOROS_SERVER_REQUEST_FILTER_CONFIG_KEY][ConfigProvider::DIACTOROS_X_FORWARDED_FILTER_CONFIG_KEY] ?? [];
+
+        if (! is_array($config) || empty($config)) {
```
I don't think we need to check `empty()`

weierophinney replied:
Psalm was complaining, IIRC. :)

weierophinney replied:
Actually... the question here is: what should we do if there is no trusted-proxies and/or trusted-headers configuration?

The assumption I made is that if they registered this factory, and then did not provide configuration, it would be as strict as possible — hence the check for `empty($config)`. Thoughts?

Ocramius replied:
If there are no such keys, don't we default to empty arrays anyway?

weierophinney replied:
It should raise an exception; it's invalid configuration. Change pushed.

Ocramius commented on src/Container/FilterUsingXForwardedHeadersFactory.php:
```diff
@@ -0,0 +1,78 @@
+        $config = $config[ConfigProvider::DIACTOROS_CONFIG_KEY][ConfigProvider::DIACTOROS_SERVER_REQUEST_FILTER_CONFIG_KEY][ConfigProvider::DIACTOROS_X_FORWARDED_FILTER_CONFIG_KEY] ?? [];
+
+        if (! is_array($config) || empty($config)) {
+            // Trust nothing!
+            return FilterUsingXForwardedHeaders::trustProxies([], []);
```
This means that the default setup of diactoros starting from this patch would be to trust no proxies: shouldn't we use the defaults of diactoros, which are BC compliant, and which come with private networks as default trusted proxies?

weierophinney replied:
Remember, this factory is _opt-in_, so the idea here would be that if you opt-in _without also providing configuration_, you're trusting nothing.

By default, no filter is provided, so it uses the Diactoros defaults (`trustReservedSubnets()`). That means it remains BC.

Ocramius replied:
Indeed, my bad! I missed this bit later on:

https://github.com/mezzio/mezzio/blob/ac9a047569b427ae79b18275651a1df1001f8fa2/src/Container/ServerRequestFactoryFactory.php#L44-L46

Ocramius commented on src/Container/FilterUsingXForwardedHeadersFactory.php:
```diff
@@ -0,0 +1,78 @@
+            return FilterUsingXForwardedHeaders::trustProxies([], []);
+        }
+
+        $proxies = array_key_exists(ConfigProvider::DIACTOROS_TRUSTED_PROXIES_CONFIG_KEY, $config)
```
`??` would also be OK here

weierophinney replied:
:+1:

Ocramius commented on src/Container/FilterUsingXForwardedHeadersFactory.php:
```diff
@@ -0,0 +1,78 @@
+            ? $config[ConfigProvider::DIACTOROS_TRUSTED_PROXIES_CONFIG_KEY]
+            : [];
+
+        if (! is_string($proxies) && ! is_array($proxies)) {
```
Broken record, but the `is_string()` branch shouldn't exist

weierophinney replied:
Done.

Ocramius commented on src/Container/FilterUsingXForwardedHeadersFactory.php:
```diff
@@ -0,0 +1,78 @@
+            return FilterUsingXForwardedHeaders::trustProxies([], []);
+        }
+
+        $proxies = is_string($proxies) ? [$proxies] : $proxies;
```
Not necessary once removing the `is_string()` branch, but `(array) $proxies` does the same

weierophinney replied:
Done.

Ocramius commented on src/Container/FilterUsingXForwardedHeadersFactory.php:
```diff
@@ -0,0 +1,78 @@
+            return FilterUsingXForwardedHeaders::trustProxies($proxies);
+        }
+
+        if (! is_array($headers)) {
+            // Invalid value; trust nothing
+            return FilterUsingXForwardedHeaders::trustProxies([], []);
+        }
```
This should throw, not use a silent success

weierophinney replied:
Done.

weierophinney committed:
docs: register DoNotFilter as an invokable, not an alias

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: invalid configuration means an exception

- Only allow arrays for the trusted proxies
- If we do not have array valus for trusted proxies or headers, it means we have INVALID CONFIGURATION, and we should raise an exception.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: resolve CS issues flagged by phpcs

Used phpcbf; mostly missing or obsolete import statements.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on src/Container/FilterUsingXForwardedHeadersFactory.php:
```diff
@@ -0,0 +1,77 @@
+        if (! is_array($headers)) {
+            // Invalid value; trust nothing
+            throw Exception\InvalidTrustedHeaderConfigurationException::forType(
+                is_object($headers) ? get_class($headers) : gettype($headers)
```
How about we give `$headers` to this exception constructor, and let it do the translation into the correct message?

weierophinney replied:
The main reason I didn't was to enforce type constraints. Using this approach allowed me to type the argument provided to the exception, which meant I didn't get any "MixedAssignment" issues flagged by Psalm.

If these are okay with you, I'll change; just wanted to provide the "why" behind this decision.

Ocramius replied:
When declaring `mixed` explicitly, it should be fine :thinking:

weierophinney replied:
Indeed - must have done something weird previously. Resolved now!

Ocramius reviewed (approved):
Besides CS issues + letting the exceptions produce the right message, this LGTM.

weierophinney committed:
refactor: move type coercion to exception constructors

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius reviewed (approved):
LGTM :ship: :rocket:

meidlinga commented:
There is a typo in the usage description. 

The class is referenced as `\Laminas\Diactoros\ServerRequestFilter\ServerRequestFilterInterface::class`, but it is named 
`Laminas\Diactoros\ServerRequestFilter\FilterServerRequestInterface::class`

see: 
https://github.com/mezzio/mezzio/blob/d5bf81f0caee6cf4b682a4bf1f8d693288bf217d/src/Container/ServerRequestFactoryFactory.php#L8
and:
https://github.com/laminas/laminas-diactoros/blob/2.25.x/src/ServerRequestFilter/FilterServerRequestInterface.php

Xerkus commented:
@meidlinga please provide PR

meidlinga commented:
@Xerkus : I provided a PR for the docs in #149 .

@weierophinney : could you change the example in the issue description from `ServerRequestFilterInterface` to `FilterServerRequestInterface`? It is linked in the release notes and I think other people would use it as a reference.

weierophinney commented:
@meidlinga Looks like somebody beat me to it. :smile:

meidlinga commented:
@weierophinney sorry then, that was to fast for me. :)
